File: zenspider/2018/19.hs
{-# LANGUAGE RecordWildCards #-} -- for R {..} -> R { a=a, b=b ... }

-- import Data.Char
import Data.Bits ((.&.), (.|.))
import Utils
--
-- import qualified Data.List as L
import qualified Data.Map as M

{-# ANN file' "HLint: ignore Defined but not used" #-}
{-# ANN main' "HLint: ignore Defined but not used" #-}

type Prim = Int -> Int -> Int
type Op   = Register -> Instruction -> Register

-- -- NOTE: I'm not sold on Reg as a separate type yet.
data Reg         = R0 | R1 | R2 | R3 | R4 | R5 deriving (Show, Eq, Ord, Enum)
data Register    = R { pc, r0, r1, r2, r3, r4, r5 :: Int } deriving (Show, Eq, Ord)
data Instruction = I { op :: Op, a, b :: Int, c :: Reg } -- deriving (Show, Eq)

newR :: Register
newR = R 0 0 0 0 0 0 0

parseInst :: [String] -> Instruction
parseInst s = I op a b (i2r c)
  where [op', a', b', c'] = s
        [a, b, c] = fmap readi [a', b', c']
        op = ops M.! op'

i2r :: Int -> Reg
i2r c =
  case c of
    0 -> R0
    1 -> R1
    2 -> R2
    3 -> R3
    4 -> R4
    5 -> R5
    _ -> error "nope!"

get :: Register -> Reg -> Int
get R {..} c =                  -- Honestly, code like this seems dumb
  case c of
    R0 -> r0
    R1 -> r1
    R2 -> r2
    R3 -> r3
    R4 -> r4
    R5 -> r5

geti :: Register -> Int -> Int
geti r c = get r (i2r c)

set :: Register -> Reg -> Int -> Register
set r c v =                     -- Same... dumb.
  case c of
    R0 -> r { r0 = v }
    R1 -> r { r1 = v }
    R2 -> r { r2 = v }
    R3 -> r { r3 = v }
    R4 -> r { r4 = v }
    R5 -> r { r5 = v }

uprr :: Register -> Instruction -> Prim -> Register
uprr r I {..} fn = set r c (geti r a `fn` geti r b)

upri :: Register -> Instruction -> Prim -> Register
upri r I {..} fn = set r c (geti r a `fn` b)

upir :: Register -> Instruction -> Prim -> Register
upir r I {..} fn = set r c (a `fn` geti r b)

addr, addi, mulr, muli, banr, bani, borr, bori,
  setr, seti, gtir, gtri, gtrr, eqir, eqri, eqrr :: Op

addr r o = uprr r o (+)      --  1 opcode, see `ins` below
addi r o = upri r o (+)      -- 13
mulr r o = uprr r o (*)      -- 15
muli r o = upri r o (*)      -- 14
banr r o = uprr r o (.&.)    --  0
bani r o = upri r o (.&.)    --  9
borr r o = uprr r o (.|.)    --  8
bori r o = upri r o (.|.)    --  5
setr r o = uprr r o const    --  3
seti r o = set r (c o) (a o) --  7
gtir r o = upir r o gti      --  6
gtri r o = upri r o gti      -- 12
gtrr r o = uprr r o gti      --  4
eqir r o = upir r o eqi      -- 10
eqri r o = upri r o eqi      --  2
eqrr r o = uprr r o eqi      -- 11

booli    :: Bool -> Int
booli b   = if b then 1 else 0
gti, eqi :: Int -> Int -> Int
gti a b   = booli $ a > b
eqi a b   = booli $ a == b

ops :: M.Map String Op
ops = M.fromList [("addr", addr),
       ("addi", addi),
       ("mulr", mulr),
       ("muli", muli),
       ("banr", banr),
       ("bani", bani),
       ("borr", borr),
       ("bori", bori),
       ("setr", setr),
       ("seti", seti),
       ("gtir", gtir),
       ("gtri", gtri),
       ("gtrr", gtrr),
       ("eqir", eqir),
       ("eqri", eqri),
       ("eqrr", eqrr)]

exec :: Register -> Reg -> [Instruction] -> Register
exec reg slot is = go reg
  where
    maxI         = length is
    updateSlot r = set r slot
    updatePC   r = r { pc = get r slot + 1 }
    go r         =
      if pc' >= maxI
        then r
        else let q   = updateSlot r pc' -- copy PC to R0
                 i   = is !! pc'        -- get instruction via pc
                 r'  = op i q i         -- execute instructon
                 r'' = updatePC r'      -- copy R0 to PC & increment PC
              in go r''
      where
        pc' = pc r

munge :: String -> (Int, [Instruction])
munge s = (ip, is)
  where
    ([_, ip'], is') = splitAt 2 $ words s
    ip              = readi ip'
    is''            = splitEvery 4 is'
    is              = fmap parseInst is''

problem1 :: String -> String
problem1 s = show $ exec reg (i2r ip) is
  where
    (ip, is) = munge s
    reg      = newR

problem2 :: String -> String
problem2 s = show $ exec reg (i2r ip) is
  where
    (ip, is) = munge s
    reg      = newR { r0 = 1 }

main :: IO ()
main = minteract [problem1, problem2]

----------------------------------------------------------------------

main' :: IO ()
main' = xinteract [problem1] file'

file' :: String
file' =
  unlines
    [ "#ip 0"
    , "seti 5 0 1"
    , "seti 6 0 2"
    , "addi 0 1 0"
    , "addr 1 2 3"
    , "setr 1 0 0"
    , "seti 8 0 4"
    , "seti 9 0 5"
    ]

File: zenspider/2018/19.swift
import Foundation

struct Reg {
    var r0: Int = 0; var r1: Int = 0; var r2: Int = 0;
    var r3: Int = 0; var r4: Int = 0; var r5: Int = 0;
    var pc: Int = 0

    subscript(idx: Int) -> Int {
        get {
            switch idx {        // ordered via actual usage data
            case 2: return r2; case 5: return r5; case 3: return r3
            case 4: return r4; case 1: return r1; case 0: return r0
            case _: fatalError("no")
            }
        }

        set(val) {
            switch idx {
            case 2: r2 = val; case 5: r5 = val; case 3: r3 = val
            case 4: r4 = val; case 1: r1 = val; case 0: r0 = val
            case _: fatalError("no")
            }
        }
    }
}

typealias Prim = (Int, Int) -> Int
typealias Op   = (inout Reg, Inst) -> Void
typealias Inst = (inout Reg) -> Void

func b2i(test: Bool) -> Int {
    return test ? 1 : 0
}

let gti = { (a: Int, b: Int) in b2i(test: a > b) }
let eqi = { (a: Int, b: Int) in b2i(test: a == b) }

func readFileLines(path: String) -> [String] {
    do {
        return try String(contentsOfFile: path).components(separatedBy: "\n")
    } catch {
        return []
    }
}

let t = true
let f = false

func wrap_inst(_ fn: @escaping Prim, _ a: Int, _ b: Int, _ c: Int, _ aReg: Bool, _ bReg: Bool?) -> Inst {
    switch (aReg, bReg) {
    case (t, t):   return { (r : inout Reg) in r[c] = fn(r[a], r[b]) }
    case (t, f):   return { (r : inout Reg) in r[c] = fn(r[a],   b ) }
    case (f, t):   return { (r : inout Reg) in r[c] = fn(  a , r[b]) }
    case (f, f):   return { (r : inout Reg) in /* no-op */           } // FIX how is this not exhaustive?!?
    case (t, nil): return { (r : inout Reg) in r[c] =    r[a]        }
    case (f, nil): return { (r : inout Reg) in r[c] =      a         }
    default:       return { (r : inout Reg) in /* no-op */           }
    }
}

func handle_op(_ words: [String]) -> Inst {
    let o = words[0]
    let a = Int(words[1]) ?? 0
    let b = Int(words[2]) ?? 0
    let c = Int(words[3]) ?? 0

    switch o {
    case "addr": return wrap_inst(+,   a, b, c, t, t)
    case "addi": return wrap_inst(+,   a, b, c, t, f)
    case "mulr": return wrap_inst(*,   a, b, c, t, t)
    case "muli": return wrap_inst(*,   a, b, c, t, f)
    case "banr": return wrap_inst(&,   a, b, c, t, t)
    case "bani": return wrap_inst(&,   a, b, c, t, f)
    case "borr": return wrap_inst(|,   a, b, c, t, t)
    case "bori": return wrap_inst(|,   a, b, c, t, f)
    case "setr": return wrap_inst(^,   a, b, c, t, nil)
    case "seti": return wrap_inst(^,   a, b, c, f, nil)
    case "gtir": return wrap_inst(gti, a, b, c, f, t)
    case "gtri": return wrap_inst(gti, a, b, c, t, f)
    case "gtrr": return wrap_inst(gti, a, b, c, t, t)
    case "eqir": return wrap_inst(eqi, a, b, c, f, t)
    case "eqri": return wrap_inst(eqi, a, b, c, t, f)
    case "eqrr": return wrap_inst(eqi, a, b, c, t, t)
    default:     return wrap_inst(^,   a, b, c, f, f) // f, f == no-op
    }
}

func parse(path: String) -> (Int, [Inst]) {
    var slot = 0
    let lines = readFileLines(path: path)

    let insts: [Inst] = lines.compactMap { str in
        let words = str.components(separatedBy: " ")

        switch words.count {
        case 2:
            slot = Int(words[1]) ?? 0
            return nil

        case 4:
            return handle_op(words)

        default:
            return nil
        }
    }

    return (slot, insts)
}

func exec(r: Reg, insts: [Inst], slot: Int) -> Reg {
    var r = r // convert let to var... seems hacky
    let start = CFAbsoluteTimeGetCurrent()
    let max = insts.count

    while r.pc < max {
        let pc = r.pc
        let op = insts[pc]

        r[slot] = pc
        op(&r)
        r.pc = r[slot] + 1
    }

    let end = CFAbsoluteTimeGetCurrent() - start
    print("Took \(end) seconds")

    return r
}

func run() {
    setbuf(__stdoutp, nil) // don't buffer output

    let start = CFAbsoluteTimeGetCurrent()
    if CommandLine.arguments.count < 2 {
        for _ in 1...100 {
            CommandLine.arguments.append("/Users/ryan/Work/git/searbsg/advent_of_code/zenspider/2018/19.txt")
        }
    }

    for path in CommandLine.arguments[1...] {
        let (slot, insts) = parse(path: path)

        print(exec(r: Reg(), insts: insts, slot: slot).r0)
    }
    let end = CFAbsoluteTimeGetCurrent() - start
    print("Total took \(end) seconds")
}

let _top: Void = run()
